### PR TITLE
Add a new metric to measure apiserver request filter latency

### PR DESCRIPTION
**Pull request by tkashem**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Measure how much time a request spends in server filter(s):    
- A: the timestamp a particular filter starts executing for a request.
- B: the timestamp the particular filter has completed and the next handler in the chain has started executing.
    
We define filter latency as `B - A`. Introduce a new metric to track filter latency `apiserver_request_filter_duration_seconds`. This measures request filter latency distribution in seconds, for each filter type.

Enable filter latency tracking for the following filters:
- apf (priority and fairness)
- authentication
- authorization
- impersonation
- audit

Query to graph the filter latency by each request filter:
```
histogram_quantile(0.99, sum(rate(apiserver_request_filter_duration_seconds_bucket[5m])) by(filter,le))
```

The following screenshot shows filter latency in action on a dev cluster for `authentication`, `authorization`, `audit` and `impersonation`:
image

The following shows for priority and fairness
image

Both graphs are scaled to milliseconds.

```release-note
A new metric `apiserver_request_filter_duration_seconds` has been introduced that 
measures request filter latency in seconds.
```

### CODE REVIEW

**Comment by tkashem** [comment]
@wojtek-t  @mm4tt I ran tests using the above server filters: https://github.com/tkashem/graceful/blob/master/priority-fairness/filter-latency/readme.md

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 1]
```diff
@@ -196,6 +196,18 @@ var (
 		[]string{"verb", "resource", "subresource"},
 	)
 
+	requestFilterLatency = compbasemetrics.NewHistogramVec(
+		&compbasemetrics.HistogramOpts{
+			Name: "apiserver_request_filter_duration",
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:201]: What units are these in? Should that go in the name? Or at least the description?

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 2]
```diff
@@ -196,6 +196,18 @@ var (
...
+			Help: "How much time the request spends in apf",
+			Buckets: []float64{0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 15.0, 20.0, 25.0, 30.0,
+				35.0, 40.0, 45.0, 50.0, 60.0, 70.0, 80.0, 90.0, 100.0, 125.0, 150.0, 175.0, 200.0, 300.0,
+				400.0, 500.0, 600.0, 700.0, 800.0, 900.0, 1000.0},
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:205]: The default settings will hold something for up to 15s.

How do you plan to represent requests that are held for the max time and then rejected?

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 3]
```diff
@@ -302,6 +315,25 @@ func UpdateInflightRequestMetrics(phase string, nonmutating, mutating int) {
...
+	scope := CleanScope(requestInfo)
+	verb := canonicalVerb(strings.ToUpper(req.Method), scope)
+	// set verbs to a bounded set of known and expected verbs
+	if !validRequestMethods.Has(verb) {
+		verb = OtherRequestMethod
+	}
+
+	ms := elapsed.Milliseconds()
+	if requestInfo.IsResourceRequest {
+		requestFilterLatency.WithLabelValues(name, cleanVerb(verb, req), requestInfo.APIGroup, requestInfo.APIVersion, requestInfo.Resource).Observe(float64(ms))
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:331]: Oh it's ms? I think metrics are usually microseconds or seconds, at least I didn't expect this.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 4]
```diff
@@ -196,6 +196,18 @@ var (
 		[]string{"verb", "resource", "subresource"},
 	)
 
+	requestFilterLatency = compbasemetrics.NewHistogramVec(
+		&compbasemetrics.HistogramOpts{
+			Name: "apiserver_request_filter_duration",
+			Help: "How much time the request spends in apf",
+			Buckets: []float64{0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 15.0, 20.0, 25.0, 30.0,
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:203]: My first thought: measure the .1-50ms region with high resolution, and then use exponential buckets up to 60s (4x the in-practice max with default flags).

File: staging/src/k8s.io/apiserver/pkg/server/filters/artificial_delayer.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	"fmt"
+	"net/http"
+	"time"
+
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+
+	"k8s.io/klog/v2"
+)
+
+// WithArtificialDelayAdder adds artificial delay to the .
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/filters/artificial_delayer.go:29]: cliffhanger

File: staging/src/k8s.io/apiserver/pkg/server/filters/filter_decorator.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+func WithFiletrStartTimestamp(parent context.Context, startTimestamp time.Time) context.Context {
+	return apirequest.WithValue(parent, filterStartTimestampKey, startTimestamp)
+}
+
+func FilterStartTimestampFrom(ctx context.Context) (time.Time, bool) {
+	info, ok := ctx.Value(filterStartTimestampKey).(time.Time)
+	return info, ok
+}
+
+func DecorateFilter(chain http.Handler, name string, longRunningRequestCheck apirequest.LongRunningRequestCheck, decorate func(h http.Handler) http.Handler) http.Handler {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/filters/filter_decorator.go:44]: I don't understand this file at all, why is it so complicated? Is this for testing?

File: staging/src/k8s.io/apiserver/pkg/server/filters/filter_decorator.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+
+	"k8s.io/klog/v2"
+)
+
+type filterStartTimestampKeyType int
+
+const filterStartTimestampKey filterStartTimestampKeyType = iota
+
+func WithFiletrStartTimestamp(parent context.Context, startTimestamp time.Time) context.Context {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/filters/filter_decorator.go:35]: speeelling :)

File: staging/src/k8s.io/apiserver/pkg/server/filters/filter_decorator.go (hunk 6 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/server/filters/filter_decorator.go:44]: I guess this is for timestamping before/after? I wouldn't use the word "decorate" for that, it's too general.

I'm not sure about this approach in general, P&F might already know the duration. I guess I'd expect to update this metric when P&F makes a decision about the request.

**Comment by tkashem** [comment]
@lavalamp I apologize, I should have made this clear. this is only for running a test. None of it is intended to be merged.

**Comment by wojtek-t** [comment]
> @lavalamp I apologize, I should have made this clear. this is only for running a test. None of it is intended to be merged.

Ohh - I thought you wanted to merge that, it's my fault then.
[BTW - I think having those kind of metrics would actually be super useful in production too. So I think we should try to have that somehow in if possible.]

**Comment by lavalamp** [comment]
Yeah I like this metric!

**Comment by tkashem** [comment]
> [BTW - I think having those kind of metrics would actually be super useful in production too. So I think we should try to have that somehow in if possible.]

Hi @wojtek-t, that sounds good to me. I will refactor, make it more readable and then open it for review.

**Comment by tkashem** [comment]
> [BTW - I think having those kind of metrics would actually be super useful in production too. So I think we should try to have that somehow in if possible.]

Hi @lavalamp @wojtek-t I have enabled filter latency for a few filters, please review when you have time.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 8]
```diff
@@ -196,6 +196,17 @@ var (
 		[]string{"verb", "resource", "subresource"},
 	)
 
+	requestFilterDuration = compbasemetrics.NewHistogramVec(
+		&compbasemetrics.HistogramOpts{
+			Name: "apiserver_request_filter_duration_seconds",
+			Help: "Request filter latency distribution in seconds, for each filter type",
+			Buckets: []float64{0.0, 0.0001, 0.0002, 0.0004, 0.0008, 0.001, 0.002, 0.004, 0.008, 0.01, 0.015, 0.02, 0.025,
```
Comment (yue9944882) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:203]: overall LGTM, except for one nit: 30+ buckets looks too many to me.. isn't 10-ish buckets here just fine?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go [hunk 9]
```diff
@@ -42,6 +43,14 @@ func WithAudit(handler http.Handler, sink audit.Sink, policy policy.Checker, lon
 	if sink == nil || policy == nil {
 		return handler
 	}
+
+	// We want to track how much time a request spends in audit filter.
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:47]: I initially though that having this kind of wrapper is overkill and preferred that to have more custom to the P&F.

But after thinking more about that and seeing this PR, I actually like the genericity of it and the fact we can use it for other filters too.
@lavalamp - does that seem reasonable to you too?

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go (hunk 8 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:203]: agree - that seems like an overkill to me.
All the filters should actually be pretty fast, so having such a bit granularity for larger values seems a bit useless. Especially for values above a second.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+type requestFilterRecordKeyType int
+
+// requestFilterRecordKey is the context key for a request filter record struct.
+const requestFilterRecordKey requestFilterRecordKeyType = iota
+
+type requestFilterRecord struct {
+	name               string
+	startedTimestamp   time.Time
+	completedTimestamp time.Time
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:37]: Do we really need completedTImestamp?
It's an additional time to keep in memory, so maybe having name and startedTime is enough?

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go (hunk 8 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: I set it to `[]float64{0.0, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.25, 0.50, 1.0, 5.0}`. Let me know if this looks good.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go (hunk 9 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:47]: Hi @lavalamp @wojtek-t, I have enabled it for `priority and fairness`, `authentication` `authorization`, `impersonation` and `audit` filters for now.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go (hunk 10 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go]: Yes, sounds good to me. I removed `completedTimestamp`

**Comment by tkashem** [comment]
@MikeSpreitzer  https://github.com/tkashem/graceful/blob/master/priority-fairness/filter-latency/readme.md

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go (hunk 9 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:47]: shouldn't we have this in the default handler chain builder func instead? The fan out would be smaller than adding it in all filters.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 11]
```diff
@@ -196,6 +196,16 @@ var (
 		[]string{"verb", "resource", "subresource"},
 	)
 
+	requestFilterDuration = compbasemetrics.NewHistogramVec(
+		&compbasemetrics.HistogramOpts{
+			Name:           "apiserver_request_filter_duration_seconds",
+			Help:           "Request filter latency distribution in seconds, for each filter type",
+			Buckets:        []float64{0.0, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.25, 0.50, 1.0, 5.0},
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:203]: Roughly exponential progression: good.
Note: the square root of 10 is about 3, not 5 or 2.

**Review by MikeSpreitzer — COMMENTED** [review]
A few overall thoughts:

I agree with https://github.com/kubernetes/kubernetes/pull/95207#discussion_r510824094

Given that `context.WithValue(..)` always pushes an association, this builds up a long context chain.  That's a bit of an anti-pattern.  How about side-effecting when possible instead of pushing a new association?

I see this is measuring the overhead on the way in but not on the way out.  Would it be better to do both?

For the API Priority and Fairness filter, the current measurement covers two rather distinct sorts of time periods: the overhead of the filter, and the time spent waiting in a queue (in the cases where that happens).  It would be good to be able to observe the two separately.  Fortunately we already have `apiserver_flowcontrol_request_wait_duration_seconds` for the latter.  If all we care about is the average then we can subtract the average wait duration from the average total to get the average filter overhead.  If we want quantiles then this is not possible, rather a more direct observation of the filter overhead is needed.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go (hunk 9 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: @sttts @MikeSpreitzer I have moved the tracking inside the `DefaultBuildHandlerChain` function now.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 12]
```diff
@@ -669,17 +670,39 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
...
+
 	failedHandler := genericapifilters.Unauthorized(c.Serializer)
 	failedHandler = genericapifilters.WithFailedAuthenticationAudit(failedHandler, c.AuditBackend, c.AuditPolicyChecker)
+
+	// We want to track how much time a request spends in authentication filter.
+	failedHandler = filterlatency.TrackCompleted(failedHandler)
+	handler = filterlatency.TrackCompleted(handler)
 	handler = genericapifilters.WithAuthentication(handler, c.Authentication.Authenticator, failedHandler, c.Authentication.APIAudiences)
+	handler = filterlatency.TrackStarted(handler, "authentication")
+
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:705]: :+1: – a little more verbose here, but nicely contained where we define the chain.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 13]
```diff
@@ -669,17 +670,39 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
...
+	// We want to track how much time a request spends in authorization filter.
+	handler := filterlatency.TrackCompleted(apiHandler)
+	handler = genericapifilters.WithAuthorization(handler, c.Authorization.Authorizer, c.Serializer)
+	handler = filterlatency.TrackStarted(handler, "authorization")
+
 	if c.FlowControl != nil {
+		// We want to track how much time a request spends in PriorityAndFairness filter.
+		handler = filterlatency.TrackCompleted(handler)
 		handler = genericfilters.WithPriorityAndFairness(handler, c.LongRunningFunc, c.FlowControl)
+		handler = filterlatency.TrackStarted(handler, "apf")
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:682]: apf? priorityandfairness ?

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 14]
```diff
@@ -669,17 +670,39 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 }
 
 func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
-	handler := genericapifilters.WithAuthorization(apiHandler, c.Authorization.Authorizer, c.Serializer)
+	// We want to track how much time a request spends in authorization filter.
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:673]: do we really need these comments?

**Comment by sttts** [comment]
/approve
Lgtm – giving others time to review 

@logicalhan @wojtek-t @deads2k

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 15]
```diff
@@ -669,17 +670,34 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
...
+		handler = filterlatency.TrackCompleted(handler)
 		handler = genericfilters.WithPriorityAndFairness(handler, c.LongRunningFunc, c.FlowControl)
+		handler = filterlatency.TrackStarted(handler, "priorityandfairness")
 	} else {
 		handler = genericfilters.WithMaxInFlightLimit(handler, c.MaxRequestsInFlight, c.MaxMutatingRequestsInFlight, c.LongRunningFunc)
 	}
+
+	handler = filterlatency.TrackCompleted(handler)
 	handler = genericapifilters.WithImpersonation(handler, c.Authorization.Authorizer, c.Serializer)
+	handler = filterlatency.TrackStarted(handler, "impersonation")
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:687]: can we factor this a little tighter?  Perhaps an `Add` method that does the wrapping or a filter around just the WithImpersonation call?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:687]: note that we need both: start and complete.

**Comment by wojtek-t** [comment]
> Lgtm – giving others time to review
> @logicalhan @wojtek-t @deads2k

LGTM - I'm leaving tagging it to others though.

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 15 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:687]: @deads2k I initially had this pattern but it looked less readable imo. I had to use a function like this to get it into one function:
```
func Tracked(handler http.Handler, name string, tracked func(h http.Handler) http.Handler) http.Handler
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go (hunk 11 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:203]: How about
```suggestion
			Buckets:        []float64{0.0, 0.0003, 0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1.0, 5.0},
```

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 15 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:687]: Yeah, either way would be OK with me.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+func trackStarted(handler http.Handler, name string, clock utilclock.PassiveClock) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		fr := &requestFilterRecord{
+			name:             name,
+			startedTimestamp: clock.Now(),
+		}
+		r = r.WithContext(withRequestFilterRecord(ctx, fr))
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:72]: Building up a long chained context is an anti-pattern.  How about the following instead?

```suggestion
		if fr := requestFilterRecordFrom(ctx); fr != nil {
			fr.name = name
			fr.startedTimestamp = clock.Now()
		} else {
			fr := &requestFilterRecord{
				name:             name,
				startedTimestamp: clock.Now(),
			}
			r = r.WithContext(withRequestFilterRecord(ctx, fr))
		}
```

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:79]: @MikeSpreitzer yeah, that sounds like a good idea! I made the change and added a test to verify.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go (hunk 11 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:203]: sounds good, I made the change.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency_test.go [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	utilclock "k8s.io/apimachinery/pkg/util/clock"
+)
+
+func TestTrackStartedWithNoFilterRecordInContext(t *testing.T) {
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency_test.go:28]: I think you got the names of this and the following test swapped.

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency_test.go]: Fixed, thanks for catching it!

**Review by MikeSpreitzer — APPROVED** [review]
/LGTM

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 18]
```diff
@@ -196,6 +196,16 @@ var (
 		[]string{"verb", "resource", "subresource"},
 	)
 
+	requestFilterDuration = compbasemetrics.NewHistogramVec(
+		&compbasemetrics.HistogramOpts{
+			Name:           "apiserver_request_filter_duration_seconds",
+			Help:           "Request filter latency distribution in seconds, for each filter type",
+			Buckets:        []float64{0.0, 0.0003, 0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1.0, 5.0},
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:203]: Having a 0.0 bucket doesn't make much sense - we should switch that to 0.0001 or sth like that.
@tkashem

**Pull request merged**